src/cmd/services/m3coordinator/downsample/id_pool_types.go

```diff
@@ -215,6 +218,25 @@ func (p *rollupIDProvider) Duplicate() ident.TagIterator {
 	return duplicate
 }
 
+func (p *rollupIDProvider) Hash() (uint32, error) {
+	if len(p.tagPairs) == 0 {
+		return 0, nil
+	}
+	if p.digest == 0 {
+		digest := adler32.New()
+		for _, t := range p.tagPairs {
```

Comment:
Instead of looping over tag pairs it might be easier to refactor ".Current()" to be ".current(index int)" and then do a for loop from zero to the end and ask for the tag at each index?

Otherwise we'll be missing the special tags we insert as part of the rollup provider and only be using the base tags for this hash?